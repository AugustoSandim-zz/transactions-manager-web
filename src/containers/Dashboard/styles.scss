@import "../../utils/style/variables.scss";

.dashboard {
  height: 100vh;
  background: rgba(72, 74, 160, 1);
  background: linear-gradient(
    to bottom,
    rgba(72, 74, 160, 1) 0%,
    rgba(72, 74, 160, 1) 49.99%,
    rgba(255, 255, 255, 1) 50%,
    rgba(255, 255, 255, 1) 100%
  );

  &__title {
    text-align: center;
    padding: 20px;
    color: white;
    font-family: "Oswald", sans-serif;
  }
  
  &__amount {
    color: white;
    display: block;
    font-size: 48px;
    padding: 0 24px; 
    
    span {
      font-size: 14px;
    }
  }

  &__container {
    display: flex;
    height: stretch;
    width: 80%;
    margin: 40% auto;
    justify-content: center;
    flex-flow: row;
  }

  &__list {
    &--empty {
      background-color: white;
      height: 100px;
      font-size: 20px;
      border-radius: 10px;
      text-align: center;
      box-shadow: 0 0 15px rgba(0, 0, 0, 0.15);
      display: flex;
      flex-flow: column;
      justify-content: center;
      width: 100%;
      margin: 0 auto;
      padding: 0 60px;
    }
  }

  &__new-transaction {
    position: absolute;
    border: none;
    background-color: $scale12;
    border-radius: 50%;
    color: white;
    width: 64px;
    height: 64px;
    padding: 20px;
    text-align: center;
    font-size: 24px;
    box-shadow: 1px 1px 4px rgba(0, 0, 0, 0.4);
    cursor: pointer;
    bottom: 0;
    margin: 24px;
    right: 0;
  }
}
