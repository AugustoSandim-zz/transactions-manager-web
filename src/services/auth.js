export const TOKEN_KEY = "@transactions-manager-Token";
export const isAuthenticated = () => localStorage.getItem(TOKEN_KEY) !== null;
export const currentUser = () => localStorage.getItem(TOKEN_KEY);

export const login = token => {
  localStorage.setItem(TOKEN_KEY, token);
};

export const logout = () => {
  localStorage.removeItem(TOKEN_KEY);
};