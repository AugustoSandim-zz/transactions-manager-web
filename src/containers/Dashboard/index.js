import React, {useState, useEffect} from 'react';
import {Link, withRouter} from 'react-router-dom';
import {currentUser} from '../../services/auth';

function Dashboard() {
  const [transactions, setTransactions] = useState([]);

  const fetchTransactions = () => {
    const list = JSON.parse(localStorage.getItem(currentUser()));
    // Order by date created
    list.sort((a, b) => new Date(b.created) - new Date(a.created));

    return setTransactions(list);
  };

  useEffect(() => {
    fetchTransactions();
  }, []);

  const calculateTotal = () => {
    if (!transactions) return 0;

    return transactions.reduce((acc, item) => {
      return (
        Number(acc) +
        Number(item.transactionValue.replace('R$', '').replace(',', '.'))
      );
    }, 0);
  };

  const renderTotalValue = () => {
    return `TOTAL R$ ${calculateTotal()
      .toFixed(2)
      .replace('.', ',')}`;
  };

  return (
    <div className="dashboard">
      <h1>Dashboard</h1>
      <strong>{renderTotalValue()}</strong>
      {transactions ? (
        transactions.map(item => (
          <li key={item.created}>
            {item.description} - {item.transactionValue}
          </li>
        ))
      ) : (
        <>
          <h3>Nenhuma transação cadastrada!</h3>
        </>
      )}
      <Link to="/nova-transacao">Nova transacao</Link>
    </div>
  );
}

export default withRouter(Dashboard);
