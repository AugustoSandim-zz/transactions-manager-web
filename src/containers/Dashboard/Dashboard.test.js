/* global expect, jest */
import React from 'react';
import {mount} from 'enzyme';
import {Route, MemoryRouter} from 'react-router-dom';
import Dashboard from './index';
import {login} from '../../services/auth';

function renderDashboard() {
  return mount(
    <MemoryRouter>
      <Route render={props => <Dashboard {...props} />} />
    </MemoryRouter>,
  );
}

beforeEach(() => {
  login('user@test.com');
});

describe('Dashboard', () => {
  it('renders without crashing', () => {
    const component = renderDashboard();

    expect(component).toMatchSnapshot();
  });

  it('renders message empty list', () => {
    const component = renderDashboard();

    expect(component.text()).toContain('Nenhuma transação cadastrada!');
  });

  it('renders value of total amount zero', () => {
    const component = renderDashboard();
    const amount = component.find('.dashboard__amount');

    expect(amount.text()).toContain('0,00');
  });

  it('renders list with card item', () => {
    global.localStorage.setItem(
      'user@test.com',
      JSON.stringify([
        {
          created: '2019-10-01T05:39:46.753Z',
          description: 'Descrição aqui',
          transactionValue: 'R$ 621,52',
        },
      ]),
    );

    const component = renderDashboard();
    component.update();

    const amount = component.find('.dashboard__amount');

    expect(component.find('.item').text()).toContain('R$ 621,52');
    expect(amount.text()).not.toContain('R$ 0,00');
  });

  it('renders total amount  correctly', () => {
    global.localStorage.setItem(
      'user@test.com',
      JSON.stringify([
        {
          created: '2019-10-01T05:39:46.753Z',
          description: 'Descrição aqui',
          transactionValue: 'R$ 621,52',
        },
        {
          created: '2019-10-02T05:39:46.753Z',
          description: 'Segunda',
          transactionValue: 'R$ 328,65',
        },
      ]),
    );

    const component = renderDashboard();
    component.update();

    const amount = component.find('.dashboard__amount');

    expect(amount.text()).toContain('R$ 950,17');
  });
});
