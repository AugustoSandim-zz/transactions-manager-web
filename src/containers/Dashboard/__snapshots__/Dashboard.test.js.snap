// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders without crashing 1`] = `ReactWrapper {}`;
